src/main/java/com/coreos/jetcd/internal/impl/LeaseImpl.java

```diff
@@ -82,19 +84,29 @@
   @Override
   public CompletableFuture<LeaseGrantResponse> grant(long ttl) {
     LeaseGrantRequest leaseGrantRequest = LeaseGrantRequest.newBuilder().setTTL(ttl).build();
-    return Util.toCompletableFuture(
+    return Util.toCompletableFutureWithRetry(
         this.stub.leaseGrant(leaseGrantRequest),
         LeaseGrantResponse::new,
+        () -> {
+          this.stub = connectionManager.newStubWithNewToken(LeaseGrpc::newFutureStub);
```

Comment:
@GuardedBy is an marker annotation only, you should then use the right synchronization block, in this case

```java
synchronized(this) {
    ...
}
```